File: python/pages/base_page.py
"""Shared behaviour for all page objects."""

from __future__ import annotations

from playwright.sync_api import Page, expect


class BasePage:
    """Base class holding the ``Page`` handle and common app-shell actions."""

    def __init__(self, page: Page) -> None:
        self.page = page
        self.menu_button = page.locator("#react-burger-menu-btn")
        self.logout_link = page.get_by_role("link", name="Logout")

    def goto(self, path: str = "/") -> None:
        self.page.goto(path, wait_until="domcontentloaded")

    def logout(self) -> None:
        self.menu_button.click()
        self.logout_link.click()
        expect(self.page.get_by_role("button", name="Login")).to_be_visible()

    @property
    def url(self) -> str:
        return self.page.url

File: typescript/src/pages/BasePage.ts
import { type Page, type Locator, expect } from '@playwright/test';

/**
 * BasePage holds behaviour shared by every page object.
 *
 * Best practices applied:
 *  - Composition over inheritance of raw selectors: subclasses expose typed
 *    {@link Locator}s, never `ElementHandle`s.
 *  - Web-first assertions everywhere (auto-waiting, no manual sleeps).
 */
export abstract class BasePage {
  protected readonly page: Page;

  /** Hamburger menu shared across the authenticated app shell. */
  readonly menuButton: Locator;
  readonly logoutLink: Locator;

  constructor(page: Page) {
    this.page = page;
    this.menuButton = page.locator('#react-burger-menu-btn');
    this.logoutLink = page.getByRole('link', { name: 'Logout' });
  }

  /** Navigate to a path relative to the configured `baseURL`. */
  async goto(path = '/'): Promise<void> {
    await this.page.goto(path, { waitUntil: 'domcontentloaded' });
  }

  /** Open the slide-out menu and log the user out. */
  async logout(): Promise<void> {
    await this.menuButton.click();
    await this.logoutLink.click();
    await expect(
      this.page.getByRole('button', { name: 'Login' }),
    ).toBeVisible();
  }

  /** Current page URL — handy for assertions and debugging. */
  url(): string {
    return this.page.url();
  }
}
